### PR TITLE
[tcat] implementation of tcat Ping command

### PR DESCRIPTION
**Pull request by canisLupus1313**
Commit implements TCAT ping command `kTlvPing`.

### CODE REVIEW

**Comment by canisLupus1313** [comment]
@EskoDijk  @arnulfrupp FYI.

File: src/core/meshcop/tcat_agent.cpp [hunk 1]
```diff
@@ -504,6 +519,24 @@ Error TcatAgent::HandleDecomission(void)
...
 }
 
+Error TcatAgent::HandlePing(const Message &aIncommingMessage,
+                            Message       &aOutgoingMessage,
+                            uint16_t       aOffset,
+                            uint16_t       aLength)
+{
+    Error           error = kErrorNone;
+    ot::ExtendedTlv extTlv;
+    VerifyOrExit(aLength <= kResponseWithStatusMaxLenght, error = kErrorParse);
```
Comment (maciejbaczmanski) [inline src/core/meshcop/tcat_agent.cpp:529]:
```suggestion
    ot::ExtendedTlv extTlv;

    VerifyOrExit(aLength <= kResponseWithStatusMaxLenght, error = kErrorParse);
```

File: tools/tcat_ble_client/cli/base_commands.py [hunk 2]
```diff
@@ -103,6 +105,38 @@ async def execute_default(self, args, context):
...
+        return 'Echo request to device.'
+
+    async def execute_default(self, args, context):
+        bless: BleStreamSecure = context['ble_sstream']
+        size = 10
+        max_payload = 512
+        if len(args) > 0:
+            size = int(args[0])
+            if size > max_payload:
+                print(f'Payload to large. Maximum supported value is {max_payload}')
```
Comment (maciejbaczmanski) [inline tools/tcat_ble_client/cli/base_commands.py:120]:
```suggestion
                print(f'Payload is too large. Maximum supported value is {max_payload}')
```

File: tools/tcat_ble_client/cli/base_commands.py [hunk 3]
```diff
@@ -103,6 +105,38 @@ async def execute_default(self, args, context):
...
 
+class PingCommand(Command):
+
+    def get_help_string(self) -> str:
+        return 'Echo request to device.'
+
+    async def execute_default(self, args, context):
+        bless: BleStreamSecure = context['ble_sstream']
+        size = 10
+        max_payload = 512
```
Comment (maciejbaczmanski) [inline tools/tcat_ble_client/cli/base_commands.py:116]: I'd suggest renaming these variables to make it more readable:
```suggestion
        payload_size = 10
        max_payload_size = 512
```

File: tools/tcat_ble_client/cli/base_commands.py [hunk 4]
```diff
@@ -103,6 +105,38 @@ async def execute_default(self, args, context):
...
+            size = int(args[0])
+            if size > max_payload:
+                print(f'Payload to large. Maximum supported value is {max_payload}')
+                return
+        to_send = token_bytes(size)
+        data = TLV(TcatTLVType.PING.value, to_send).to_bytes_extended()
+        elapsed_time = time()
+        response = await bless.send_with_resp(data)
+        print(response)
+        elapsed_time = time() - elapsed_time
```
Comment (maciejbaczmanski) [inline tools/tcat_ble_client/cli/base_commands.py:127]:
```suggestion
        elapsed_time = time() - elapsed_time
        print(response)
```
to measure only sending process, without `print()` time

File: src/core/meshcop/tcat_agent.cpp [hunk 5]
```diff
@@ -413,6 +413,21 @@ Error TcatAgent::HandleSingleTlv(const Message &aIncommingMessage, Message &aOut
         case kTlvDecommission:
             error = HandleDecomission();
             break;
+        case kTlvPing:
+            if (tlv.IsExtended())
```
Comment (EskoDijk) [inline src/core/meshcop/tcat_agent.cpp:417]: both base and extended format TLVs are allowed for all commands - see spec 13.9.  A sender should use the extended format only if the payload doesn't fit in the base format.

File: src/core/meshcop/tcat_agent.cpp [hunk 6]
```diff
@@ -504,6 +519,24 @@ Error TcatAgent::HandleDecomission(void)
     return error;
 }
 
+Error TcatAgent::HandlePing(const Message &aIncommingMessage,
```
Comment (EskoDijk) [inline src/core/meshcop/tcat_agent.cpp:522]:
```suggestion
Error TcatAgent::HandlePing(const Message &aIncomingMessage,
```

File: src/core/meshcop/tcat_agent.hpp [hunk 7]
```diff
@@ -361,7 +362,8 @@ class TcatAgent : public InstanceLocator, private NonCopyable
     bool         CanProcessTlv(uint8_t aTlvType) const;
     CommandClass GetCommandClass(uint8_t aTlvType) const;
 
-    static constexpr uint16_t kJoinerUdpPort = OPENTHREAD_CONFIG_JOINER_UDP_PORT;
+    static constexpr uint16_t kJoinerUdpPort               = OPENTHREAD_CONFIG_JOINER_UDP_PORT;
+    static constexpr uint16_t kResponseWithStatusMaxLenght = 512;
```
Comment (EskoDijk) [inline src/core/meshcop/tcat_agent.hpp:366]:
```suggestion
    static constexpr uint16_t kPingPayloadMaxLength = 512;
```

suggest a rename here - it's about the max length of the payload in a "Ping" message/command.
(Also swapped t/h at the end)  In general a payload may be even longer, for example when requesting a Commissioner's certificate. But for ping it's indeed limited to 512.

File: src/core/meshcop/tcat_agent.hpp [hunk 8]
```diff
@@ -353,6 +353,7 @@ class TcatAgent : public InstanceLocator, private NonCopyable
     Error HandleSingleTlv(const Message &aIncommingMessage, Message &aOutgoingMessage);
     Error HandleSetActiveOperationalDataset(const Message &aIncommingMessage, uint16_t aOffset, uint16_t aLength);
     Error HandleDecomission(void);
+    Error HandlePing(const Message &aIncommingMessage, Message &aOutgoingMessage, uint16_t aOffset, uint16_t aLength);
```
Comment (EskoDijk) [inline src/core/meshcop/tcat_agent.hpp:356]:
```suggestion
    Error HandlePing(const Message &aIncomingMessage, Message &aOutgoingMessage, uint16_t aOffset, uint16_t aLength);
```

File: tools/tcat_ble_client/cli/base_commands.py [hunk 9]
```diff
@@ -103,6 +105,38 @@ async def execute_default(self, args, context):
         return CommandResultTLV(tlv_response)
 
 
+class PingCommand(Command):
+
+    def get_help_string(self) -> str:
+        return 'Echo request to device.'
```
Comment (EskoDijk) [inline tools/tcat_ble_client/cli/base_commands.py:111]:
```suggestion
        return 'Send echo request to TCAT device.'
```

File: tools/tcat_ble_client/cli/base_commands.py (hunk 2 above)
Comment (EskoDijk) [inline tools/tcat_ble_client/cli/base_commands.py:120]:
```suggestion
                print(f'Payload size too large. Maximum supported value is {max_payload}')
```

File: tools/tcat_ble_client/cli/base_commands.py [hunk 10]
```diff
@@ -103,6 +105,38 @@ async def execute_default(self, args, context):
...
+        bless: BleStreamSecure = context['ble_sstream']
+        size = 10
+        max_payload = 512
+        if len(args) > 0:
+            size = int(args[0])
+            if size > max_payload:
+                print(f'Payload to large. Maximum supported value is {max_payload}')
+                return
+        to_send = token_bytes(size)
+        data = TLV(TcatTLVType.PING.value, to_send).to_bytes_extended()
```
Comment (EskoDijk) [inline tools/tcat_ble_client/cli/base_commands.py:123]: could use the extended TLV type only if the payload size is > 254 -> that may help to test the TCAT device as DUT, so we can check it supports both base and extended TLV types.   Ideally the TLV class would encode automatically to base or extended type depending on need.

**Review by EskoDijk — CHANGES_REQUESTED** [review]
Thanks, I've added some comments / change requests!

File: tools/tcat_ble_client/cli/base_commands.py (hunk 4 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: fixed.

File: tools/tcat_ble_client/cli/base_commands.py (hunk 3 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: fixed.

File: src/core/meshcop/tcat_agent.cpp (hunk 1 above)
Comment (canisLupus1313) [reply src/core/meshcop/tcat_agent.cpp]: fixed.

File: src/core/meshcop/tcat_agent.cpp (hunk 5 above)
Comment (canisLupus1313) [reply src/core/meshcop/tcat_agent.cpp]: I changed the code to handle both possibilities.

File: src/core/meshcop/tcat_agent.cpp (hunk 6 above)
Comment (canisLupus1313) [reply src/core/meshcop/tcat_agent.cpp]: It's strange that the code spell didn't catch those, I've fixed all of the occurrences.

File: src/core/meshcop/tcat_agent.hpp (hunk 7 above)
Comment (canisLupus1313) [reply src/core/meshcop/tcat_agent.hpp]: fixed.

File: tools/tcat_ble_client/cli/base_commands.py (hunk 9 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: fixed.

File: src/core/meshcop/tcat_agent.hpp (hunk 8 above)
Comment (canisLupus1313) [reply src/core/meshcop/tcat_agent.hpp]: fixed.

File: tools/tcat_ble_client/cli/base_commands.py (hunk 2 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: fixed.

File: tools/tcat_ble_client/cli/base_commands.py (hunk 10 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: I should work fine now.

File: tools/tcat_ble_client/cli/base_commands.py [hunk 11]
```diff
@@ -103,6 +105,37 @@ async def execute_default(self, args, context):
...
+        return 'Send echo request to TCAT device.'
+
+    async def execute_default(self, args, context):
+        bless: BleStreamSecure = context['ble_sstream']
+        payload_size = 10
+        max_payload = 512
+        if len(args) > 0:
+            payload_size = int(args[0])
+            if payload_size > max_payload:
+                print(f'Payload size to large. Maximum supported value is {max_payload}')
```
Comment (maciejbaczmanski) [inline tools/tcat_ble_client/cli/base_commands.py:120]:
```suggestion
                print(f'Payload size too large. Maximum supported value is {max_payload}')
```

**Review by maciejbaczmanski — APPROVED** [review]
LGTM, one nit

File: tools/tcat_ble_client/cli/base_commands.py (hunk 11 above)
Comment (canisLupus1313) [reply tools/tcat_ble_client/cli/base_commands.py]: fixed.

**Review by EskoDijk — APPROVED** [review]
Thanks, LGTM!

**Comment by canisLupus1313** [comment]
@abtink @jwhui Can I count on your review as well?

File: src/core/meshcop/tcat_agent.cpp [hunk 12]
```diff
@@ -504,6 +511,35 @@ Error TcatAgent::HandleDecomission(void)
...
+                            Message       &aOutgoingMessage,
+                            uint16_t       aOffset,
+                            uint16_t       aLength)
+{
+    Error           error = kErrorNone;
+    ot::ExtendedTlv extTlv;
+    ot::Tlv         tlv;
+
+    VerifyOrExit(aLength <= kPingPayloadMaxLength, error = kErrorParse);
+    if (aLength > 254)
```
Comment (abtink) [inline src/core/meshcop/tcat_agent.cpp:524]:
```suggestion
    VerifyOrExit(aLength <= kPingPayloadMaxLength, error = kErrorParse);
    
    if (aLength > ot::Tlv::kBaseTlvMaxLength)
```

**Review by abtink — APPROVED** [review]
Look good. Thanks.
Couple of smaller suggestions:

**Review by jwhui — APPROVED** [review]

**Pull request merged**